packages/flutter/lib/src/widgets/text_formatter.dart

```diff
@@ -0,0 +1,98 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart';
+
+/// A [TextInputFormatter] can be optionally injected into an [EditableText] 
+/// to provide as-you-type validation and formatting of the text being edited.
+/// 
+/// Text modification should only be applied when text is being committed by the
+/// IME and not on text under composition (i.e. when 
+/// [TextEditingValue.composing] is collapsed). 
+/// 
+/// A concrete [BlacklistingTextInputFormatter] which removes blacklisted 
+/// characters upon edit commit is provided. 
+/// 
+/// To create custom formatters, extend the [TextInputFormatter] class and 
+/// implement the [formatEditUpdate] method.
+/// 
+/// See also:
+///  
+///  * [EditableText] on which the formatting apply.
+///  * [BlacklistingTextInputFormatter], a provided formatter for blacklisting
+///    characters.
+abstract class TextInputFormatter {
+  /// Called when text is being typed or cut/copy/pasted in the [EditableText].
+  /// 
+  /// You can override the resulting text based on the previous text value and
+  /// the incoming new text value.
+  TextEditingValue formatEditUpdate(
+    TextEditingValue oldValue, 
+    TextEditingValue newValue
+  );
+}
+
+/// A [TextInputFormatter] that prevents the insertion of blacklisted.
+/// 
+/// Instances of blacklisted characters found in the new [TextEditingValue]s 
+/// will be replaced with the [replacementString] which defaults to ``.
+/// 
+/// It attempts to preserve the existing [TextEditingValue.selection] to 
+/// sensible values. 
+/// 
+/// See also:
+///  
+///  * [TextInputFormatter].
+class BlacklistingTextInputFormatter extends TextInputFormatter {
+  BlacklistingTextInputFormatter(
+    this.blacklistedPattern, 
+    {this.replacementString: ''}) : 
+    assert(blacklistedPattern != null);
+
+  /// A [Pattern] to match and replace incoming [TextEditingValue]s.
+  final Pattern blacklistedPattern;
+  /// String used to replace found patterns.
+  final String replacementString;
+
+  @override
+  TextEditingValue formatEditUpdate(
+    TextEditingValue oldValue, 
+    TextEditingValue newValue
+  ) {
+    // Don't apply formatting to text still under composition.
```

Comment:
I was pedantic with the wording. But pragmatically, I don't think it's possible to enter a newline while composing.